build-system/pr-check/utils.js

```diff
@@ -233,6 +261,35 @@ function timedExecOrDie(cmd, fileName = 'utils.js') {
   stopTimer(cmd, fileName, startTime);
 }
 
+/**
+ * Executes the command and check unpermitted DNS requests in the interim. In
+ * case if failure, it terminates.
+ * @param {string} cmd
+ * @param {string} fileName
+ */
+async function timedExecOrDieWithDnsMonitor(cmd, fileName = 'utils.js') {
+  return startDnsMonitor_().then(dnsMonitor => {
+    timedExecOrDie(cmd, fileName);
+    if (dnsMonitor) {
+      dnsMonitor.stdout.pipe(process.stdout);
+      dnsMonitor.stderr.pipe(process.stderr);
+      dnsMonitor.on('exit', code => {
+        if (code) {
+          process.exit(1);
+        }
+        resolver();
+      });
+      dnsMonitor.kill('SIGTERM');
+      let resolver;
+      return new Promise(resolverIn => {
+        resolver = resolverIn;
+      });
+    } else {
+      return Promise.resolve();
+    }
+  });
+}
```

Comment:
I'd like to request a fundamental change to the way the `gulp dns-monitor` task works, which will make this function unnecessary.

Right now, the responsibility of running `gulp dns-monitor` in the background falls on the caller, and different callers will have to independently figure out how to start a background daemon.

Instead, I suggest you make `gulp dns-monitor` start as a daemon process on its own. That way, you can run something like this:
```
gulp dns-monitor --start
gulp e2e
gulp dns-monitor --stop
```

There is an example for this behavior in the `gulp nailgun` task:

https://github.com/ampproject/amphtml/blob/1b2413852b83d7b4ce7fd403df9fcf54321fea94/build-system/tasks/nailgun.js#L186-L187

Reply:
I've used `gulp nailgun` as the blueprint when writing this! For this dns-monitor, I feel the biggest difference is that its return code (0 or 1) would dictate whether the Travis test should fail. We could do that in theory, but it would require a new way for doing inter-process communication.

Reply:
As discussed offline, my main objection is that most of the code in `timedExecOrDieWithDnsMonitor` is specific to the dns monitor and doesn't belong in `utils.js`.